### PR TITLE
[sparse] Mask padded entries in _coo_fromdense_jvp and _csr_fromdense_jvp (fixes #40624)

### LINKED ISSUE #40624: jax.experimental.sparse.coo_fromdense gives incorrect second derivative when nse includes padded entries
### Description

`jax.experimental.sparse.coo_fromdense` appears to produce an incorrect second-order reverse-mode derivative when `nse` is larger than the number of true nonzero entries.

In the example below, the dense input has three true nonzero entries, while `nse=5`. The extra `out.data` entries are zero padding and should remain constant zero. Therefore, the local scalar function should be:

```
(t + 1)**2 + t**2 + (t**2 + 2)**2
```

At `t = 0.02`, the expected second derivative is `12.0048`. However, JAX returns `16.0048`.

### Minimal reproducible example

```
#!/usr/bin/env python3
import os
import sys
import math

os.environ.setdefault("CUDA_VISIBLE_DEVICES", "")
os.environ.setdefault("OMP_NUM_THREADS", "1")
os.environ.setdefault("JAX_PLATFORMS", "cpu")

try:
    import jax
    import jax.numpy as jnp
    import jax.experimental.sparse
except ImportError as e:
    print(f"missing dep: {e}")
    sys.exit(2)

jax.config.update("jax_enable_x64", True)

def target(t):
    z = jnp.zeros_like(t)
    a = jnp.stack((t + 1, z, -t, z, z, t * t + 2)).reshape((2, 3)).T
    out = jax.experimental.sparse.coo_fromdense(
        a,
        nse=5,
        index_dtype=jnp.int32,
    )
    return jnp.sum(out.data ** 2)

x = jnp.asarray(0.02, dtype=jnp.float64)
expected = 12.0048

actual = float(jax.grad(jax.grad(target))(x))

print("actual:", actual)
print("expected:", expected)

if not (math.isfinite(actual) and abs(actual - expected) <= 1e-12):
    sys.exit(0)
sys.exit(1)
```

### Actual result

```
actual: 16.0048
expected: 12.0048
```

### Expected result

```
12.0048
```

The expected value follows from differentiating the scalar function formed by the true nonzero entries:

```
(t + 1)**2 + t**2 + (t**2 + 2)**2
```

The padded entries introduced by `nse=5` should remain constant zero and should not contribute to the second derivative. JAX instead returns `16.0048`, which suggests that padded sparse entries may be leaking into the second-order autodiff result.

### System info (python version, jaxlib version, accelerator, etc.)

jax: 0.10.1
jaxlib: 0.10.1
numpy: 2.4.6
python: 3.13.13
platform: Ubuntu 24.04.3 LTS, Linux x86_64
accelerator: CPU backend

Note: GPU is present, but CUDA-enabled jaxlib is not installed, so JAX falls back to CPU.

### PR DESCRIPTION
**Pull request by jdymitarai**
Fixes #40624.

### Background & Problem
When converting a dense array to sparse COO or CSR format (`coo_fromdense` or `csr_fromdense`) where `nse` exceeds the actual number of nonzero entries (`nnz`), `_coo_fromdense_impl` and `_csr_fromdense_impl` properly set padded elements in `data` to 0.

However, in `_coo_fromdense_jvp` and `_csr_fromdense_jvp`:
- `_coo_extract(row, col, Mdot)` and `_csr_extract(indices, indptr, Mdot)` extracted tangent values at all `nse` indices without masking out the padding.
- Because padded index entries default to `(0, 0)` in `COO` (or clamped index boundaries in `CSR`), tangents of the dense matrix at those coordinates leaked into the padded entries of `data_dot`.
- When computing second-order derivatives (such as `grad(grad(target))` in #40624) or differentiating linear functions over `out.data`, these phantom tangents leaked into the derivative, returning incorrect values (e.g. `16.0048` instead of `12.0048`).

### Solution
1. In `_coo_fromdense_jvp` and `_csr_fromdense_jvp`, mask tangents with `true_nonzeros = jnp.arange(nse) < (M != 0).sum()`, ensuring padded tangent entries are always `0`.
2. Parameterize `test_coo_fromdense_ad` and `test_csr_fromdense_ad` to test `nse_padding=[0, 3]`.
3. Add a dedicated regression test `test_coo_fromdense_padded_second_derivative_issue40624`.

### CODE REVIEW

**Comment by jdymitarai** [comment]
Closing for now to keep the PR queue focused. Thanks!
